lib/Index/Index.cpp

```diff
@@ -153,6 +154,20 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
 
   bool IsModuleFile = false;
   bool isSystemModule = false;
+
+  struct ContainerStackEntry {
+    Decl *const VisitedDecl;
+    // These vector capacities represent the number of declarations that may
+    // be associated with the key type. In most cases there will be only one; a
+    // function or single named pattern binding. However in the case of an
+    // expanded tuple pattern binding, there will be multiple.
+    SmallVector<Decl *, 6> ActiveDecls{};
+    std::map<Expr *, SmallVector<Decl *, 6>> ExprDecls{};
+    std::map<TypeRepr *, SmallVector<Decl *, 6>> TypeReprDecls{};
+    std::map<Pattern *, SmallVector<Decl *, 6>> PatternDecls{};
```

Comment:
I suppose there's no harm in marking the pointer const, but I probably wouldn't bother (the vector and maps are still mutable).

Also, given that's there's normally only a single `Decl *`, it seems a bit wasteful to use a `SmallVector<... 6>` as the value in the maps. What do you think about something like:
```
struct ContainerStackEntry {
  Decl *VisitedDecl;
  llvm::PointerUnion<Decl *, TuplePattern *> Active;
  // Don't both storing anything if there's no corresponding decls (ie. all AnyPatterns)
  DenseMap<void *, llvm::PointerUnion<Decl *, TuplePattern *>> DeclMap;
};
```
?

`addContainedByRelationIfContained` could then handle either the `Decl*` or `TuplePattern*`.

And speaking of patterns, do we actually need to keep the map decls for a pattern? The corresponding TypeRepr/Expr will have a mapping and custom attributes can only apply for a single var, so that would already be covered by the active case.

Reply:
> Also, given that's there's normally only a single Decl *, it seems a bit wasteful to use a SmallVector<... 6> as the value in the maps. What do you think about something like:

I agree it's not ideal, though I don't believe your proposed alternative would work. There can be more than one active decl, and they're not always associated with a `TuplePattern`, e.g:
```
let (a, b) = someFunc() // both a and b are active so that each contains someFunc() references.
```
```
func someFunc() {
    let _: Int = 1 // The active container for Int is someFunc(), as the AnyPattern has no VarDecl.
}
```

I did however condense the three maps into one using `void *` (I wasn't sure if using `void *` was considered good c++ or not 😄).

> And speaking of patterns, do we actually need to keep the map decls for a pattern?

When we've a single var in the binding (or an AnyPattern), we just associate the Pattern and decl and ignore the TypeReps and Exprs; whatever types and expressions that follow all need to be contained by the single var anyway. This greatly simplifies [this logic here](https://github.com/apple/swift/pull/38883/commits/49e854dd3a56a02a8f84c0fddfa79f94dfd4ade1#diff-187a51395f73cbba56fa55021f185ae2a88b30e4a7306de23800aba8a235c634R551-R589), which can focus solely on unpacking expanded tuple bindings.

Reply:
> I agree it's not ideal, though I don't believe your proposed alternative would work. There can be more than one active decl, and they're not always associated with a TuplePattern, e.g:

When is there more than one decl but those aren't all from a TuplePattern? In your example both `(a, b)` are part of the TuplePattern. Then for the AnyPattern case that would just be the `someFunc` Decl (both able to be stored as a `PointerUnion<Decl *, TuplePattern *>`. Ie. my suggestion was essentially that we either associate the single Decl or the TuplePattern.

> When we've a single var in the binding (or an AnyPattern), we just associate the Pattern and decl and ignore the TypeReps and Exprs; whatever types and expressions that follow all need to be contained by the single var anyway

I'm still missing why associating the Pattern is required there - can't we just set the active decl to the single var in that case?

Reply:
I got it working using a `PointerUnion`, please see the latest commit. I think your suggestion of using `TuplePattern` was throwing me off a little, as there's actually other patterns at play here. However we can just use `Pattern` instead, and avoid the need for an extra type in the union:

```
let (c, d) = someFunc() // TuplePattern
let (a, b): (Int, Int) = ... // TypedPattern
```

> I'm still missing why associating the Pattern is required there - can't we just set the active decl to the single var in that case?

Currently when we encounter a `TypedPattern` with a `TupleTypeRepr` we can iterate through the type elements and assume we've a matching number of var decls. However given the binding `let foo: (Int, Int) = (1, 1)` that's not a safe assumption.

So instead of explicitly handling this scenario (along with `AnyPattern` and maybe others) it's simpler and more future proof to just associate the active decl with the `Pattern` when we encounter a single `VarDecl`.

Reply:
So there are other `Patterns`, but we're handling them by creating the mappings in `identifyPatternBindingContainerDecls`. For the actual mappings themselves we only care about either a single `Decl`, or a `TuplePattern` that contains all the `NamedPatterns`. The `TuplePattern` *has* to contain all NamedPatterns` as otherwise we would have mapped to `Decl` (or the current container).

This also means that `ActiveDecls` can just be `Active` and a `PointerUnion<Decl *, TuplePattern *>` instead. 

To hopefully make that a little more concrete, the AST for `let (a, b): (Int, String) = (expr1, expr2)` will look something like:
```
PDB
  TypedPattern
    TuplePattern
      NamedPattern(a)
      NamedPattern(b)
    TupleTypeRepr
      IdentTypeRepr(Int)
      IdentTypeRepr(String)
  TupleExpr
    expr1
    expr2
```

So in this case we'd want the following mappings:
  - Int/expr1 -> Decl(a)
  - String/expr2 -> Decl(b)

The `TypedPattern` IMO should always map to `nullptr`, which would allow us to use the same stack for `getParent` (ie. a mapping of `nullptr` == use container, which you currently do for the `AnyPattern` case).

Following from this, looking at `let a: (Int, Int) = (1, 1)` the AST would be:
```
PDB
  TypedPattern
    NamedPattern
  TupleTypeRepr
    ...
  TupleExpr
    ...
```

If the sub pattern of the `TypedPattern` is not a `TuplePattern`, then we can just create a mapping:
  - TypedPattern -> nullptr
  - Type/Expr -> Decl (unless it's `AnyPattern`, in which case -> nullptr)

`Active` would start as `Decl(a)` for the single variable case in order to handle property wrappers. For the tuple case it could start as the container.


I also just noticed that this doesn't currently handle nested `TuplePattern`, ie.
```
let ((a, b), c): ((Int, String), Double) = ((1, ""), 2.0)
```

Reply:
> So there are other Patterns, but we're handling them by creating the mappings in identifyPatternBindingContainerDecls. For the actual mappings themselves we only care about either a single Decl, or a TuplePattern that contains all the NamedPatterns. The TuplePattern has to contain all NamedPatterns as otherwise we would have mapped to Decl (or the current container).

Sorry I don't think I explained myself fully in my last comment. I do understand that for multiple vars they will always be within a `TuplePattern`. However, that point I wanted to make is that using `PointerUnion<void *, Pattern *>` instead allows us to avoid some extra walking in `identifyPatternBindingContainerDecls`.

```
if (auto TP = dyn_cast<TypedPattern>(P)) {
    auto TR = TP->getTypeRepr();

    if (auto TTR = dyn_cast<TupleTypeRepr>(TR)) {
        for (auto n : range(TTR->getNumElements())) {
        auto ET = TTR->getElementType(n);
        setupContainerStackEntryWithVarDecl(Entry, n, ET, MaybeVarDecls);
        }
    } else {
        // Non-tuple type, use all VarDecls in this pattern.
        Entry.DeclMap[TR] = {false, P};
    }
}
```

In the `else` branch here, we can just assign whatever pattern `P` is, it doesn't actually matter. To only use `TuplePattern` in the `PointerUnion` we'd need to check if `P->getSubPattern()` is a `TuplePattern`. It most likely is, however the point is that we don't _need_ to know, simply using `P` is sufficient because all we need is to be able to call `collectVariables` later. Swift is a rapidly evolving language, so my rationale is that the fewer the type checks we need to make, the better; it makes the code less susceptible to breakages when/if new patterns are introduced.

The same applies when handling the initializer:

```
if (auto E = PBD->getInit(i)) {
    if (auto TE = dyn_cast<TupleExpr>(E)) {
        for (auto n : range(TE->getNumElements())) {
        auto E = TE->getElement(n);
        setupContainerStackEntryWithVarDecl(Entry, n, E, MaybeVarDecls);
        }
    } else {
        // Non-tuple initializer, use all VarDecls in this pattern.
        Entry.DeclMap[E] = {false, P};
    }
}
```

That's two places that'd we need to perform some extra walking, when it really isn't necessary. I'm not beholden to using `PointerUnion<void *, Pattern *>`, but if you really think we should constrain it to `TuplePattern`, I'm happy to change it.

> The TypedPattern IMO should always map to nullptr, which would allow us to use the same stack for getParent (ie. a mapping of nullptr == use container, which you currently do for the AnyPattern case).

I'm not sure I follow. For the `PointerUnion` to hold `TypedPattern`, we couldn't then use `TuplePattern`, which is what I think you've been advocating for? Are you also proposing that we merge `EntitiesStack` and `ContainerStack`?

> This also means that ActiveDecls can just be Active and a PointerUnion<Decl *, TuplePattern *> instead.

I did consider making this change, however it'd require that we call `TuplePattern::collectVariables` every time the active container is requested, vs. just once when we walk to the corresponding `TypeRepr`/`Expr` (in `activateContainerDecls`). Given the following var, that would be a lot of extra calls to `collectVariables`:

```
let (a, b): (Int, ((Int, Double, Float, String))) = (intValue, (intValue, doubleValue, floatValue, stringValue))
```

So it boils down to a trade-off between memory usage vs. execution workload. I'm not sure which we should optimize for?

Reply:
> I also just noticed that this doesn't currently handle nested TuplePattern, ie.

This is now handled by the latest commit.

Reply:
> That's two places that'd we need to perform some extra walking, when it really isn't necessary

Fair enough, `Pattern *` is fine with me.

> So it boils down to a trade-off between memory usage vs. execution workload. I'm not sure which we should optimize for?

`collectVariables` isn't a particularly intensive function, so I'd be happy running it each time.

> This is now handled by the latest commit.

Thanks, I added some comments inline